runtime/lua/vim/lsp/protocol.lua

```diff
@@ -776,10 +774,49 @@ function protocol.make_client_capabilities()
   }
 end
 
+local if_nil = vim.F.if_nil
 --- Creates a normalized object describing LSP server capabilities.
 ---@param server_capabilities table Table of capabilities supported by the server
 ---@return table Normalized table of capabilities
 function protocol.resolve_capabilities(server_capabilities)
+  local TextDocumentSyncKind = protocol.TextDocumentSyncKind
+  local textDocumentSync = server_capabilities.textDocumentSync
+  if textDocumentSync == nil then
+    -- Defaults if omitted.
+    server_capabilities.textDocumentSync = {
+      openClose = false,
+      change = TextDocumentSyncKind.None,
+      willSave = false,
+      willSaveWaitUntil = false,
+      save = {
+        includeText = false,
+      }
+    }
+  elseif type(textDocumentSync) == 'number' then
+    -- Backwards compatibility
+    if not TextDocumentSyncKind[textDocumentSync] then
+      return nil, "Invalid server TextDocumentSyncKind for textDocumentSync"
+    end
+    server_capabilities.textDocumentSync = {
+      openClose = true,
+      change = textDocumentSync,
+      willSave = false,
+      willSaveWaitUntil = false,
+      save = {
+        includeText = false,
+      }
+    }
+  elseif type(textDocumentSync) ~= 'table' then
+    return nil, string.format("Invalid type for textDocumentSync: %q", type(textDocumentSync))
+  end
```

Comment:
should this assert that certain keys exist? maybe it should use `vim.validate` instead, and let the other branches return early.